Grant should be invalidated on the granter queue

pvollan opened the pull request:
#### 7516bd41dd6f4d02ab5e85bd29c71eda6dd6e5c6
<pre>
Grant should be invalidated on the granter queue
<a href="https://bugs.webkit.org/show_bug.cgi?id=272398">https://bugs.webkit.org/show_bug.cgi?id=272398</a>
<a href="https://rdar.apple.com/125984025">rdar://125984025</a>

Reviewed by Chris Dumez.

Grant should be invalidated on the granter queue and not on the main thread.

* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.h:
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
(WebKit::ExtensionCapabilityGrant::operator=): Deleted.
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
(WebKit::finalizeGrant):

Canonical link: <a href="https://commits.webkit.org/277260@main">https://commits.webkit.org/277260@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/132c75d6c04384a3f3376493df3eea86e8215dc5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47145 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26323 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49788 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49828 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43194 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31629 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23781 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38403 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47726 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23805 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40633 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19711 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21218 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5189 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42192 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51704 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22170 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18542 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45697 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23446 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44705 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24229 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6627 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23164 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -127,6 +127,13 @@ static bool finalizeGrant(const String& environmentIdentifier, AuxiliaryProcessP
+        auto& existingGrant = iterator->value;
+        ASSERT(!existingGrant.isValid());
+        if (existingGrant.isValid()) {
+            granterQueue().dispatchSync([&existingGrant]() mutable {
```
Why do we need a dispatchSync() here? Can't we move the grant to the lambda and invalidate it async?

pvollan replied:
That's a very good point, will fix!

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -127,6 +127,13 @@ static bool finalizeGrant(const String& environmentIdentifier, AuxiliaryProcessP
+                existingGrant.invalidate();
+            });
+        }
         iterator->value = WTFMove(grant);
```
Could use `existingGrant = WTFMove(grant);`

pvollan replied:
Fixed!

cdumez reviewed (approved):
LGTM

aestes commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -127,7 +127,15 @@ static bool finalizeGrant(const String& environmentIdentifier, AuxiliaryProcessP
+        ASSERT(!existingGrant.isValid());
+        if (existingGrant.isValid()) {
+            GRANTER_RELEASE_LOG_ERROR(environmentIdentifier, "grant not expected to be valid");
+            granterQueue().dispatch([existingGrant = WTFMove(existingGrant)]() mutable {
```
You should call `ExtensionCapabilityGranter::invalidateGrants` here.

pvollan replied:
Ah, good point, addressed in latest patch.

Thanks for reviewing!

aestes commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -127,7 +127,16 @@ static bool finalizeGrant(const String& environmentIdentifier, AuxiliaryProcessP
+            Vector<ExtensionCapabilityGrant> grantsToInvalidate;
+            grantsToInvalidate.reserveInitialCapacity(1);
+            grantsToInvalidate.append(WTFMove(existingGrant));
+            granter.invalidateGrants(WTFMove(grantsToInvalidate));
```
`Vector<ExtensionCapabilityGrant>::from(WTFMove(existingGrant))`?

pvollan replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
```diff
@@ -127,7 +127,13 @@ static bool finalizeGrant(const String& environmentIdentifier, AuxiliaryProcessP
+        ASSERT(!existingGrant.isValid());
+        if (existingGrant.isValid()) {
+            GRANTER_RELEASE_LOG_ERROR(environmentIdentifier, "grant not expected to be valid");
+            granter.invalidateGrants(Vector<ExtensionCapabilityGrant>::from(WTFMove(existingGrant)));
```
Wouldn't this build?
```
granter.invalidateGrants({ WTFMove(existingGrant) });
```

pvollan replied:
Ah, good point, I get the following error when trying this:

```
Copy constructor is implicitly deleted because 'ExtensionCapabilityGrant' has a user-declared move constructor
```

pvollan committed:
Grant should be invalidated on the granter queue
https://bugs.webkit.org/show_bug.cgi?id=272398
rdar://125984025

Reviewed by Chris Dumez.

Grant should be invalidated on the granter queue and not on the main thread.

* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.h:
* Source/WebKit/Platform/cocoa/ExtensionCapabilityGrant.mm:
(WebKit::ExtensionCapabilityGrant::operator=): Deleted.
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.mm:
(WebKit::finalizeGrant):

Canonical link: https://commits.webkit.org/277260@main
